src/ops/arithmetic.rs

```diff
@@ -0,0 +1,1698 @@
+    ///   2 dimensions.  The matrix dimensions for the matrix product are the last
+    ///   two dimensions of each input.
+    /// - All but the last two dimensions of each input are broadcast with one another using
+    ///   standard <doc:broadcasting>.
+    ///
+    /// # Example
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1, 2, 3, 4], &[2, 2]);
+    /// let b = Array::from_slice(&[-5.0, 37.5, 4., 7., 1., 0.], &[2, 3]);
+    ///
+    /// // produces a [2, 3] result
+    /// let mut c = a.matmul_device(&b, Default::default());
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to multiply
+    /// - stream: stream or device to evaluate on
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check that the inputs are valid for matrix multiplication.
+    #[default_device]
+    pub unsafe fn matmul_device_unchecked(&self, other: &Array, stream: StreamOrDevice) -> Array {
+        unsafe {
+            Array::from_ptr(mlx_sys::mlx_matmul(
+                self.c_array,
+                other.c_array,
+                stream.as_ptr(),
+            ))
+        }
+    }
+
+    /// Matrix multiplication returning an error if inputs are not valid.
+    ///
+    /// Perform the (possibly batched) matrix multiplication of two arrays. This function supports
+    /// broadcasting for arrays with more than two dimensions.
+    ///
+    /// - If the first array is 1-D then a 1 is prepended to its shape to make it
+    ///   a matrix. Similarly, if the second array is 1-D then a 1 is appended to its
+    ///   shape to make it a matrix. In either case the singleton dimension is removed
+    ///   from the result.
+    /// - A batched matrix multiplication is performed if the arrays have more than
+    ///   2 dimensions.  The matrix dimensions for the matrix product are the last
+    ///   two dimensions of each input.
+    /// - All but the last two dimensions of each input are broadcast with one another using
+    ///   standard <doc:broadcasting>.
+    ///
+    /// # Example
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1, 2, 3, 4], &[2, 2]);
+    /// let b = Array::from_slice(&[-5.0, 37.5, 4., 7., 1., 0.], &[2, 3]);
+    ///
+    /// // produces a [2, 3] result
+    /// let mut c = a.matmul_device(&b, Default::default());
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to multiply
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn try_matmul_device(
+        &self,
+        other: &Array,
+        stream: StreamOrDevice,
+    ) -> Result<Array, OperationError> {
+        if self.ndim() == 0 || other.ndim() == 0 {
+            return Err(OperationError::WrongInput(
+                "Got 0 dimension input. Inputs must have at least one dimension.".to_string(),
+            )
+            .into());
+        }
+
+        // if we have a 1D array, it will be reshaped to 2D
```

Comment:
Maybe we could do `[i32; 2]` instead to avoid heap alloc for something that has only 2 element?

Reply:
problem is other side is `&[i32]` didn't know how to make these two types play with each other but I've now gone with a different approach that avoids the heap alloc